README.md
# Q Applet: Weather - USA

Displays a 4-day weather forecast for cities located in the USA.

[GitHub repository](https://github.com/daskeyboard/daskeyboard-applet--weather-usa)

## Example

Displays 4-day weather forecast on a Das Keyboard Q Series.
Each color corresponds to a weather change:  
clear or sunny (yellow), cloudy (purple), rainy (blue), storm (red), snow (white).

![Weather Forecast on a Das Keybaord Q](assets/image_keys.png "Q Weather Forecast")
![Weather Forecast on a Das Keybaord Q](assets/image_legend.png "Q Weather Forecast")

## Changelog

[CHANGELOG.MD](CHANGELOG.md)

## Installation

Requires a Das Keyboard Q Series: www.daskeyboard.com

Installation, configuration and uninstallation of applets is done within the Q Desktop application (<https://www.daskeyboard.com/q>)

## Running tests

- `yarn test`

## Contributions

Pull requests welcome.

## Copyright / License

Copyright 2014 - 2019 Das Keyboard / Metadot Corp.

Licensed under the GNU General Public License Version 2.0 (or later);
you may not use this work except in compliance with the License.
You may obtain a copy of the License in the LICENSE file, or at:

   <http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Changelog Q Applet Weather USA

## Version 1.0.12 - Jan 27, 2023

Updated processing of zones json.

## Version 1.0.11 - July 28, 2020

Updated processing of zones json. Was broken in previous version due to service changes.

## Version 1.0.10 - May 10, 2019

In the configuration, get the cities sorted by state.
Removed internet connection error signal.

## Version 1.0.9 - May 07, 2019

Changed search controltype.

## Version 1.0.8 - April 30, 2019

Handled internet connection error.

## Version 1.0.7 - January 11, 2019

Priority sunny on showers and rain removed.

## Version 1.0.6 - December 06, 2018

New images and new titles added (weather-forecast become weather-usa).
README and Package.json updated.
Search issue fixed.

## Version 1.0.0 - October 30, 2018

First release


test/test.js
const assert = require('assert');
const t = require('../index');
const zoneId = 'TXZ211';
const zoneName = 'Austin, TX';

describe('evaluateForecast', function () {
  it('can evaluate percentage', function () {
    let result = t.evaluateForecast('A 20 percent chance of showers and thunderstorms. Partly cloudy, with a low around 60. East southeast wind 3 to 5 mph.');
    assert.equal(result.percent, 20);
  });
  it('can evaluate showers and thunderstorms', function () {
    let result = t.evaluateForecast('A 40 percent chance of showers and thunderstorms. Partly sunny, with a high near 84. South wind 5 to 7 mph.');
    assert.ok(result.storm, "Did not detect storm.");
    assert.ok(result.shower, "Did not detect shower.");
    assert.ok(!result.snow, "False positive on snow.");
  });
  it('can evaluate cloudy', function () {
    let result = t.evaluateForecast('Cloudy, with a high near 81.');
    assert.ok(result.cloudy, "Did not detect cloudy.");
    assert.ok(!result.snow, "False positive on snow.");
    assert.ok(!result.storm, "False positive on storm.");
  });
  it('can evaluate sunny', function () {
    let result = t.evaluateForecast('Sunny, with a high near 81.');
    assert.ok(result.sunny, "Did not detect sunny.");
    assert.ok(!result.snow, "False positive on snow.");
    assert.ok(!result.storm, "False positive on storm.");
  });
  it('can evaluate sunny', function () {
    let result = t.evaluateForecast('Clear, with a low around 57.');
    assert.ok(result.clear, "Did not detect clear.");
    assert.ok(!result.sunny, "False positive sunny.");
    assert.ok(!result.snow, "False positive on snow.");
    assert.ok(!result.storm, "False positive on storm.");
  });
  it('evaluates rain and percent chance', function () {
    let result = t.evaluateForecast('Showers and isolated thunderstorms. ' +
      'Locally heavy rainfall possible. Cooler. Lows in the upper 450s. ' +
      'North winds 10 to 20 mph. Chance of rain near 100 percent.');

    assert.ok(!result.clear, "False positive clear.");
    assert.ok(!result.sunny, "False positive sunny.");
    assert.ok(!result.snow, "False positive on snow.");
    assert.ok(result.storm, "Did not detect storm.");
    assert.ok(result.shower, "Did not detect shower.");
    assert.ok(result.percent, "Did not detect percent.");
    assert.ok(result.isLikely, "Did not detect likeliness");
  })
});

describe('getForecast', function () {
  it('can get a forecast', function () {
    t.getForecast(zoneId).then((body) => {
      console.log("Got forecast: " + JSON.stringify(body));
      const periods = body.periods;
      assert.ok(periods, "Did not get valid periods.");
      assert(periods.length > 10, "Did not get enough periods.");
    });
  });
})

describe('generateServiceHeaders', function () {
  it('can generate a header object', function () {
    const test = t.generateServiceHeaders();
    assert.ok(test);
  });
  it('generates unique service headers', function () {
    const test1 = t.generateServiceHeaders();
    const test2 = t.generateServiceHeaders();
    console.log("Service headers: " + JSON.stringify(test1));
    console.log("Service headers: " + JSON.stringify(test2));
    assert.notDeepEqual(test1, test2);
  });
});


describe('generateText', function () {
  it('can generate meaningful text', function () {
    const periods = [{
        name: 'Rest of Today',
        detailedForecast: 'Sunny'
      },
      {
        name: 'Tuesday',
        detailedForecast: '10% chance of rain'
      },
      {
        name: 'Wednesday',
        detailedForecast: '90% chance of showers'
      },
      {
        name: 'Thursday',
        detailedForecast: 'Scattered clouds'
      },
    ]
    const text = t.generateText(periods);
    const lines = text.split("\n");
    assert(lines.length == periods.length, "Incorrect number of forecast lines");
  });

  it('can handle line feeds', function () {
    const periods = [{
        name: 'Rest of Today',
        detailedForecast: "Lots\nof\nlines"
      },
      {
        name: 'Tuesday',
        detailedForecast: "Even\nmore\n\n\nlines"
      }
    ];
    const text = t.generateText(periods);
    const lines = text.split("\n");
    assert(lines.length == periods.length, "Incorrect number of forecast lines");
  });

});

describe('Observation', function () {
  it('can prioritize storm', function () {
    let test = new t.Observation({
      storm: true,
      sunny: true,
      percent: t.Observation.LIKELY_THRESHOLD
    });
    assert.equal(test.prioritize(), t.FORECASTS.STORM);
  });
  it('can prioritize shower', function () {
    let test = new t.Observation({
      shower: true,
      sunny: true,
      percent: t.Observation.LIKELY_THRESHOLD
    });
    assert.equal(test.prioritize(), t.FORECASTS.SHOWER);
  });
  // it('can prioritize sunny with unlikely shower', function () {
  //   let test = new t.Observation({
  //     shower: true,
  //     sunny: true,
  //     percent: t.Observation.LIKELY_THRESHOLD - 1
  //   });
  //   assert.equal(test.prioritize(), t.FORECASTS.SUNNY);
  // });
  it('can prioritize cloudy', function () {
    let test = new t.Observation({
      cloudy: true,
      sunny: true
    });
    assert.equal(test.prioritize(), t.FORECASTS.CLOUDY);
  });
  it('can prioritize sunny', function () {
    let test = new t.Observation({
      sunny: true
    });
    assert.equal(test.prioritize(), t.FORECASTS.SUNNY);
  });
  it('can prioritize clear', function () {
    let test = new t.Observation({
      clear: true
    });
    assert.equal(test.prioritize(), t.FORECASTS.CLEAR);
  });
  it('defaults to clear', function () {
    let test = new t.Observation({});
    assert.equal(test.prioritize(), t.FORECASTS.CLEAR);
  });
});

describe('WeatherForecast', function () {
  it('#run()', async function () {
    return buildApp().then(app => {
      app.run().then((signal) => {
        console.log(JSON.stringify(signal));
        assert.ok(signal);
        assert(signal.message.includes(zoneName));
      });
    })
  });

  it('#options()', async function () {
    return buildApp().then(app => {
      app.options('zoneId').then((options) => {
        assert.ok(options);
        assert(options.length > 1, 'Selections did not have an array of values.');
        assert(options.length > 100, 'Selections did not have enough values.');
        const option = options[0];
        assert.ok(option.key);
        assert.ok(option.value);
      })
    })
  });

  it('#options(fieldId, search)', async function () {
    return buildApp().then(app => {
      app.options('zoneId', 'texas').then((options) => {
        assert.ok(options);
        assert(options.length > 1, 'Selections did not have an array of values.');
        assert(options.length < 100, 'Search did not filter values');

        for (let option of options) {
          assert.ok(option.key);
          assert.ok(option.value);
          assert(option.value.toLowerCase().includes('texas'));
        }
      })
    })
  });

  it('#options(fieldId, search)', async function () {
    return buildApp().then(app => {
      app.options('zoneId', 'New').then((options) => {
        assert.ok(options);
        assert(options.length > 0, 'Selections did not have an array of values.');
        assert(options.length < 100, 'Search did not filter values');

        assert.ok(options.filter(option => {
          return option.value.includes('New York')
        }));
      })
    })
  });

  it('#options(fieldId, search)', async function () {
    return buildApp().then(app => {
      app.options('zoneId', 'new').then((options) => {
        assert.ok(options);
        assert(options.length > 0, 'Selections did not have an array of values.');
        assert(options.length < 100, 'Search did not filter values');

        assert.ok(options.filter(option => {
          return option.value.includes('New York')
        }));
      })
    })
  });
})

async function buildApp() {
  let app = new t.WeatherForecast();
  app.config = {
    zoneId: zoneId,
    zoneId_LABEL: zoneName,
    geometry: {
      width: 4,
      height: 1,
    }
  };

  return app;
}

index.js
const q = require('daskeyboard-applet');
const fs = require('fs');
const request = require('request-promise');
const logger = q.logger;
const apiUrl = "https://api.weather.gov";

var zones = null;

const COLORS = Object.freeze({
  CLEAR: '#FFFF00',
  CLOUDY: '#FF00FF',
  SHOWER: '#0000FF',
  SNOW: '#FFFFFF',
  STORM: '#FF0000',
  SUNNY: '#FFFF00'
})


const FORECASTS = Object.freeze({
  CLEAR: 'CLEAR',
  CLOUDY: 'CLOUDY',
  SHOWER: 'SHOWER',
  SNOW: 'SNOW',
  STORM: 'STORM',
  SUNNY: 'SUNNY'
});

class Observation {
  constructor({
    clear = false,
    cloudy = false,
    shower = false,
    snow = false,
    storm = false,
    sunny = false,
    percent = 0
  }) {
    this.clear = clear;
    this.cloudy = cloudy;
    this.shower = shower;
    this.snow = snow;
    this.storm = storm;
    this.sunny = sunny;
    this.percent = percent;
  }

  isLikely() {
    return (this.percent && this.percent >= Observation.LIKELY_THRESHOLD);
  }

  isClear() {
    return this.clear;
  }

  isCloudy() {
    return this.cloudy;
  }

  isShower() {
    return this.shower;
  }

  isSnow() {
    return this.snow && this.isLikely();
  }

  isStorm() {
    return this.storm && this.isLikely();
  }

  isSunny() {
    return this.sunny;
  }

  prioritize() {
    if (this.isSnow()) {
      return FORECASTS.SNOW;
    } else if (this.isStorm()) {
      return FORECASTS.STORM;
    } else if (this.isShower()) {
      return FORECASTS.SHOWER;
    } else if (this.isCloudy()) {
      return FORECASTS.CLOUDY;
    } else if (this.isSunny()) {
      return FORECASTS.SUNNY;
    } else {
      return FORECASTS.CLEAR;
    }
  }
}

Observation.LIKELY_THRESHOLD = 20;
const percentChanceExpression = /(\d+) percent/;


/**
 * Evaluate a forecast string for specific features
 * @param {string} forecastText 
 */
function evaluateForecast(forecastText) {
  const forecast = forecastText.toLowerCase();
  const percentMatches = percentChanceExpression.exec(forecast);

  return new Observation({
    clear: forecast.includes('clear'),
    cloudy: forecast.includes('cloudy'),
    shower: ( forecast.includes('showers') || forecast.includes('rain') ),
    snow: forecast.includes('snow'),
    storm: forecast.includes('storm'),
    sunny: forecast.includes('sunny'),
    percent: (percentMatches && percentMatches.length > 1) ?
      percentMatches[1] : '0'
  });
}

// this is a work-around to a bug in the API that seems to send stale forecasts
function generateServiceHeaders() {
  return {
    'User-Agent': 'request(q-applet-weather)',
    'Accept': `application/geo+json, application/qawf-${new Date().getTime() + '' + Math.round(Math.random() * 10000)}`,
  };
}

async function getForecast(zoneId) {
  const url = apiUrl + `/zones/ZFP/${zoneId}/forecast`;
  logger.info("Getting forecast via URL: " + url);
  return request.get({
    url: url,
    headers: generateServiceHeaders(),
    json: true
  }).then(body => {
    const periods = body.properties.periods;
    if (periods) {
      return body;
    } else {
      throw new Error("No periods returned.");
    }
  }).catch((error) => {
    logger.error(`Error when trying to getForecast: ${error}`);
    throw new Error(`Error when trying to getForecast: ${error}`);
  })
}

function generateText(periods) {
  const forecasts = [];
  for (let i = 0; i < periods.length; i += 1) {
    let text = periods[i].detailedForecast.trim();
    text = text.replace(/\n/g, " ");
    text = text.replace(/\s+/g, ' ');
    text = `<div>${text}</div>`;
    forecasts.push(`<em>${periods[i].name}:</em> ${text}`);
  }

  return forecasts.join("\n");
}

// Used for sorting search results when searching through zones.
// The state will be placed first in the value used for sorting.
// This will make it easier to find results for a specific state.
function getSortValue(zone) {
  var stateZone = zone.toUpperCase().split(", ") // ignore upper and lowercase
  return stateZone.reverse().join(" ")
}


class WeatherForecast extends q.DesktopApp {
  constructor() {
    super();
    this.zoneName = null;
    // run every 30 min
    this.pollingInterval = 30 * 60 * 1000;
  }

  async options(fieldId, search) {
    if (zones) {
      logger.info("Sending preloaded zones");
      return this.processZones(zones, search);
    } else if (fs.existsSync('./zones.json')) {
      logger.info('Loading zones from a file.');
      zones = require('./zones.json');
      return this.processZones(zones, search);
    } else {
      logger.info("Retrieving zones via API...");
      return request.get({
        url: apiUrl + '/zones?type=forecast',
        headers: generateServiceHeaders(),
        json: true
      }).then(body => {
        zones = body;
        return this.processZones(zones, search);
      }).catch((error) => {
        logger.error("Caught error:", error);
      })
    }
  }


  async applyConfig() {
    this.zoneName = this.config.zoneId_LABEL;
  }

  async getZoneName() {
    if (this.config.zoneId_LABEL) {
      return this.config.zoneId_LABEL;
    } else if (this.zoneName) {
      return this.zoneName;
    } else {
      // we save the zoneId's corresponding zone name to persistent storage so
      // that we can include it in the forecast.
      const zoneId = this.config.zoneId;
      if (zoneId) {
        const zoneInfo = this.store.get('zoneInfo');
        logger.info("My saved zoneInfo is: " + JSON.stringify(zoneInfo));

        if (zoneInfo && zoneInfo.id === zoneId) {
          this.zoneName = zoneInfo.name;
          logger.info("Retrieved zoneName: " + this.zoneName);
          return this.zoneName;
        } else {
          logger.info('Saved zone is inconsistent with configured zone. ' +
            'Retrieving from service...');
          // store the new zone name in my configuration
          const options = await this.options();
          logger.info("Checking for matching zoneId: " + zoneId);
          for (let option of options) {
            if (option.key === zoneId) {
              this.zoneName = option.value;
              logger.info("My zone name is: " + this.zoneName);

              this.store.put('zoneInfo', {
                id: zoneId,
                name: this.zoneName
              });
              break;
            }
          }
          logger.info("Finished checking for matching zoneId.");

          if (this.zoneName) {
            return this.zoneName;
          } else {
            throw new Error("Could not find zone with ID: " + zoneId);
          }
        }
      } else {
        return null;
      }
    }
  }

  /**
   * Process a zones JSON to an options list
   * @param {*} zones
   * @param {String} search 
   */
  async processZones(zones, search) {
    if (search != null) {
      search = search.trim().toLowerCase();
    }

    logger.info("Processing zones JSON");
    const options = [];
    for (let feature of zones.features) {
      if (feature.properties.type === 'public') {
        const key = feature.properties.id;
        let value = feature.properties.name;

        if (!search || value.toLowerCase().includes(search)) {
          if (feature.properties.state) {
            value = value + ', ' + feature.properties.state;
          }

          options.push({
            key: key,
            value: value
          });
        }
      }
    }

    // We sort by alphabetical order first by state and then the region/city
    let optionsSorted = options.sort(function (a, b) {
      var nameA = getSortValue(a.value)
      var nameB = getSortValue(b.value)
      if (nameA < nameB) {
        return -1;
      }
      if (nameA > nameB) {
        return 1;
      }

      // names must be equal
      return 0;
    });

    // return options[value].sort();
    return optionsSorted;

  }

  async run() {
    logger.info("Weather USA running.");
    const zoneId = this.config.zoneId;
    const zoneName = await this.getZoneName();

    if (zoneId) {
      logger.info("My zone ID is  : " + zoneId);
      logger.info("My zone name is: " + zoneName);

      return getForecast(zoneId).then(body => {
        const periods = body.properties.periods || [];
        const updated = body.properties.updated;
        const width = this.geometry.width || 4;

        logger.info("Forecast was updated on " + updated);
        logger.info(`Forecast contains ${periods.length} periods.`);
        logger.info("My width is: " + width);
        const points = [];
        const forecastPeriods = [];
        if (periods.length > 0) {
          logger.info("Got forecast: " + zoneId);
          for (let i = 0; i < width; i += 1) {
            // we skip every other one because we get a daily and nightly
            // forecast for each day
            if (periods.length > i * 2) {
              const period = periods[i * 2];
              forecastPeriods.push(period);
              const observation = evaluateForecast(period.detailedForecast);
              const forecastValue = observation.prioritize();
              const color = COLORS[forecastValue];
              points.push(new q.Point(color));
            }
          }

          const signal = new q.Signal({
            points: [points],
            name: `${zoneName}`,
            message: `<div><b>Weather Forecast for ${zoneName}:</b></div>` +
              generateText(forecastPeriods)
          });
          logger.info('Sending signal: ' + JSON.stringify(signal));
          return signal;
        } else {
          logger.info("No forecast for zone: " + zoneId);
          return null;
        }
      }).catch((error) => {
        logger.error(`Error while getting forecast data: ${error}`);
          if(`${error.message}`.includes("getaddrinfo")){
            // Commented in order to do not send this kind of notification. Boring to close the signal.
            // return q.Signal.error(
            //   'The Weather forecast USA service returned an error. <b>Please check your internet connection</b>.'
            // );
          }else{
            return q.Signal.error([`The Weather forecast USA service returned an error. Detail: ${error}`]);
          }
      })
    } else {
      logger.info("No zoneId configured.");
      return null;
    }
  }
}


module.exports = {
  FORECASTS: FORECASTS,
  Observation: Observation,
  WeatherForecast: WeatherForecast,
  evaluateForecast: evaluateForecast,
  generateServiceHeaders: generateServiceHeaders,
  getForecast: getForecast,
  generateText: generateText
}

const applet = new WeatherForecast();
